### ISSUE TITLE
Explore again the 'babylon' ast exploring with 'recast' as the parser / generator.

### ISSUE THREAD

**Issue by michalwarda** [elpassion/factorx#12]

**Comment by michalwarda** [comment]
To find all searched expressions:
```
const position = { start: 20, end: 20 }

export default function (babel) {
  const { types: t } = babel;
  
  return {
    visitor: {
      Expression(path) {
        if (path.node.start <= position.start && path.node.end >= position.end) {
	        console.log(path.node);
        }
      }
    }
  };
}
```

**Comment by michalwarda** [comment]
Extract variable to same scope:
```
const position = { start: 0, end: 1 }
const varName = 'abc'
const extractedNodes = []

export default function (babel) {
  const { types: t } = babel;
  
  return {
    visitor: {
      Expression(path) {
        const { node } = path;
        if (!node.visited && node.start === position.start && node.end === position.end) {
          node.visited = true
          const id = path.scope.generateUidIdentifierBasedOnNode(node.id);
  		  path.scope.push({ id, init: path.node });
  		  path.replaceWith(t.identifier(id.name))
          path.scope.rename(id.name, varName)
        }
      }
    }
  };
}
```

**Comment by michalwarda** [comment]
Find valid scopes to extract an identifier.
```
const position = { start: 75, end: 76 }
export default function (babel) {
  const { types: t } = babel;
  
  return {
    visitor: {
      Expression(path) {
        const { node } = path;
        console.log(node)
        if (!node.visited && node.start === position.start && node.end === position.end && t.isIdentifier(path.node)) {
          const scopes = [];
          let scope = path.scope;
          while (!scope.hasOwnBinding(node.name)) {
            scopes.push(scope)
            scope = path.scope.parent
          }
          scopes.push(scope)
        }
      }
    }
  };
}
```

**Comment by michalwarda** [comment]
This was closed by #19
